doc/designs/disable-stale-users.md

```diff
@@ -0,0 +1,291 @@
+# Disable Stale Users
+
+**DESIGN STAGE**
+
+## Overview
+
+Some regulations call for accounts that haven't been used for a specific
+amount of time be automatically disabled in the identity management system.
+
+FreeIPA does not keep track of successful authentication events.
+The krbLastSuccessfulAuth attribute is not updated by default since
+https://pagure.io/freeipa/issue/5313 for reliability reasons.
+Therefore distinguishing between stale and active accounts cannot be done with
+existing logon data.
+
+
+## Possible approaches
+
+### Environments with mandatory, periodic password changes
+
+The problem is easy to solve in environments where user passwords must be
+changed regularly. A user account whose associated password is past its due
+change date can be considered stale and therefore be disabled by the system.
+In this case, a grace period must be provided between password expiration and
+the actual account locking event.
+Total writes to LDAP using this approach due to the disable stale users tooling
+are minimal (only nsAccountLock). However, password-change related writes are
+needed every 90 day (by default).
+Each password change updates the krbLastPwdChange, krbPasswordExpiration,
+userPassword, krbPrincipalKey, and krbExtraData attributes.
+
+The Disable Stale Users (DSU) tooling proposes to make sure no non-locked
+account has a password expired since N days, N being 7 by default.
+
+### Environments with no mandatory password changes
+
+For environments where user passwords never expire, a new mechanism must be
+added to FreeIPA plugins ipa-kdb and ipa-lockout to account for each user
+account authentication events. Since updating a LDAP user account attribute on
+each authentication event must be avoided as the update frequency could easily
+be high, (see https://pagure.io/freeipa/issue/5313 for background
+information), a coarse update mechanism is proposed.
+
+An attribute containing a timestamp is added to each user:
+coarseSuccessfulAuthTime. The timestamp contained therein is not precise.
+It will be added at account creation time (set to current date).
+The timestamp is updated by IPA plugins whenever a successful
+authentication event happens, but only if the previous timestamp is
+sufficiently in the past.
+
+Three ipaConfigString attributes must be added to the cn=ipaConfig schema:
+* the first one to activate the mechanism (default: False):
+  DSU:RecordSuccessfulAuth
+* the second one contains the fixed time period (default: 28 days):
+  DSU:RecordSucessfulAuth-TimeInterval
+* the third one contains a random time period to be added (default: 7 days):
+  DSU:RecordSucessfulAuth-RandomizedTimeInterval
```

Comment:
We'd need to see whether a dashed attribute will work as an IPA option, I don't think off the top of my head it will as python will interpret it as subtraction.

Reply:
I think RandomizedTimeInterval needs to be expanded on as it is fixed at 7 by default but random?

Reply:
I removed the dash, for many reasons it was a bad idea.
I'm expanding the paragraph.